File: packages/api/src/schemas/evaluation.py
"""Pydantic schemas for the evaluation endpoints."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

from .truth import TruthPayload


class EvalQuestion(BaseModel):
    """A single evaluation question with optional expected answer."""

    question: str = Field(..., min_length=1)
    expected_answer: str | None = None
    expected_chunks: list[str] | None = Field(
        default=None, description='Expected source chunks, e.g. ["report.pdf:3", "guide.pdf"].'
    )
    truth: TruthPayload | None = None


class EvalRunCreate(BaseModel):
    """Request to create a new evaluation run."""

    model_name: str = Field(..., min_length=1)
    questions: list[str | EvalQuestion] = Field(..., min_length=1, max_length=100)
    question_set_id: int | None = None
    profile_id: str | None = None


class EvalResultResponse(BaseModel):
    """Response for a single evaluation result."""

    id: int
    question: str
    expected_answer: str | None = None
    answer: str | None = None
    contexts: str | None = None
    latency_ms: float | None = None
    relevancy_score: float | None = None
    groundedness_score: float | None = None
    context_precision_score: float | None = None
    context_relevancy_score: float | None = None
    completeness_score: float | None = None
    correctness_score: float | None = None
    compliance_accuracy_score: float | None = None
    abstention_score: float | None = None
    is_hallucination: bool | None = None
    chunk_alignment_score: float | None = None
    coverage_gaps: dict | None = None
    deterministic_checks: list[dict] | None = None
    truth: TruthPayload | None = None
    verdict: str | None = None
    fail_reasons: list[str] | None = None
    total_tokens: int | None = None
    error_message: str | None = None


class EvalRunResponse(BaseModel):
    """Response for an evaluation run."""

    id: int
    model_name: str
    question_set_name: str | None = None
    status: Literal["pending", "running", "completed", "failed", "cancelled"]
    total_questions: int
    completed_questions: int
    avg_latency_ms: float | None = None
    avg_relevancy: float | None = None
    avg_groundedness: float | None = None
    avg_context_precision: float | None = None
    avg_context_relevancy: float | None = None
    avg_completeness: float | None = None
    avg_correctness: float | None = None
    avg_compliance_accuracy: float | None = None
    avg_abstention: float | None = None
    hallucination_rate: float | None = None
    avg_chunk_alignment: float | None = None
    profile_id: str | None = None
    profile_version: str | None = None
    judge_model_name: str | None = None
    synthesis_model_name: str | None = None
    retrieval_config: dict | None = None
    corpus_snapshot: dict | None = None
    overall_verdict: str | None = None
    pass_count: int | None = None
    fail_count: int | None = None
    review_count: int | None = None
    total_tokens: int | None = None
    error_message: str | None = None
    created_at: datetime | None = None
    completed_at: datetime | None = None

    model_config = {"from_attributes": True}


class EvalRunDetailResponse(EvalRunResponse):
    """Evaluation run with individual results."""

    results: list[EvalResultResponse] = []


class EvalRunCreateResponse(BaseModel):
    """Response after creating an evaluation run."""

    eval_run_id: int
    model_name: str
    status: str
    total_questions: int
    message: str


class EvalRunRerun(BaseModel):
    """Request to re-run an evaluation with a different model."""

    model_name: str = Field(..., min_length=1)


class ComparisonMetric(BaseModel):
    """A single metric compared across two runs."""

    metric: str
    run_a: float | None = None
    run_b: float | None = None
    winner: str | None = None


class QuestionComparison(BaseModel):
    """Side-by-side comparison of a single question across two runs."""

    question: str
    expected_answer: str | None = None
    run_a: EvalResultResponse | None = None
    run_b: EvalResultResponse | None = None


class ComparisonDecision(BaseModel):
    """Backend-computed comparison verdict with disqualification gates."""

    winner: str | None = None  # "run_a" | "run_b" | "tie"
    winner_name: str | None = None
    decision_status: str = "inconclusive"  # "decisive" | "marginal" | "inconclusive"
    reason_codes: list[str] = Field(default_factory=list)
    summary: str = ""
    risk_flags: list[str] = Field(default_factory=list)
    disqualified: dict[str, list[str]] = Field(default_factory=dict)


class ComparisonWarning(BaseModel):
    """Warning about comparison precondition mismatch."""

    code: str
    message: str


class ComparisonResponse(BaseModel):
    """Side-by-side comparison of two evaluation runs."""

    run_a: EvalRunResponse
    run_b: EvalRunResponse
    metrics: list[ComparisonMetric] = []
    questions: list[QuestionComparison] = []
    decision: ComparisonDecision | None = None
    warnings: list[ComparisonWarning] = Field(default_factory=list)


class SynthesizeRequest(BaseModel):
    """Request to auto-generate evaluation questions from documents."""

    document_ids: list[int] | None = Field(
        default=None, description="Document IDs to generate from. None = all documents."
    )
    max_questions: int = Field(default=3, ge=1, le=50)
    profile_id: str | None = Field(
        default=None, description="Profile to use for domain-specific question generation."
    )


class SynthesizedQuestion(BaseModel):
    """A single auto-generated question with expected answer."""

    question: str
    expected_answer: str | None = None
    truth: TruthPayload | None = None


class SynthesizeResponse(BaseModel):
    """Response with auto-generated evaluation questions."""

    questions: list[SynthesizedQuestion] = []
    count: int = 0

File: packages/ui/src/schemas/evaluation.ts

import { z } from 'zod';

export const EvalRunCreateResponseSchema = z.object({
    eval_run_id: z.number(),
    model_name: z.string(),
    status: z.string(),
    total_questions: z.number(),
    message: z.string(),
});

export const CoverageGapsSchema = z.object({
    concepts: z.array(z.string()),
    covered: z.array(z.string()),
    missing: z.array(z.string()),
    coverage_ratio: z.number(),
    retrieval_failures: z.array(z.string()).optional(),
    generation_failures: z.array(z.string()).optional(),
});

export const DeterministicCheckSchema = z.object({
    check_name: z.string(),
    passed: z.boolean(),
    detail: z.string().optional(),
});

export const AnswerTruthSchema = z.object({
    required_concepts: z.array(z.string()),
    abstention_expected: z.boolean().optional(),
});

export const RetrievalTruthSchema = z.object({
    required_documents: z.array(z.string()),
    expected_chunk_refs: z.array(z.string()),
    supporting_documents: z.array(z.string()).optional().default([]),
    supporting_chunk_refs: z.array(z.string()).optional().default([]),
    evidence_mode: z.string(),
});

export const TruthMetadataSchema = z.object({
    truth_schema_version: z.string().optional(),
    concept_extraction_version: z.string().optional(),
    evidence_alignment_version: z.string().optional(),
    generated_by_model: z.string(),
    generated_at: z.string(),
    source_chunk_ids: z.array(z.number()),
});

export const TruthPayloadSchema = z.object({
    answer_truth: AnswerTruthSchema,
    retrieval_truth: RetrievalTruthSchema,
    metadata: TruthMetadataSchema,
});

export const EvalResultSchema = z.object({
    id: z.number(),
    question: z.string(),
    expected_answer: z.string().nullable().optional(),
    answer: z.string().nullable().optional(),
    contexts: z.string().nullable().optional(),
    latency_ms: z.number().nullable().optional(),
    relevancy_score: z.number().nullable().optional(),
    groundedness_score: z.number().nullable().optional(),
    context_precision_score: z.number().nullable().optional(),
    context_relevancy_score: z.number().nullable().optional(),
    completeness_score: z.number().nullable().optional(),
    correctness_score: z.number().nullable().optional(),
    compliance_accuracy_score: z.number().nullable().optional(),
    abstention_score: z.number().nullable().optional(),
    is_hallucination: z.boolean().nullable().optional(),
    chunk_alignment_score: z.number().nullable().optional(),
    coverage_gaps: CoverageGapsSchema.nullable().optional(),
    deterministic_checks: z.array(DeterministicCheckSchema).nullable().optional(),
    truth: TruthPayloadSchema.nullable().optional(),
    verdict: z.string().nullable().optional(),
    fail_reasons: z.array(z.string()).nullable().optional(),
    total_tokens: z.number().nullable().optional(),
    error_message: z.string().nullable().optional(),
});

export const EvalRunSchema = z.object({
    id: z.number(),
    model_name: z.string(),
    question_set_name: z.string().nullable().optional(),
    status: z.string(),
    total_questions: z.number(),
    completed_questions: z.number(),
    avg_latency_ms: z.number().nullable().optional(),
    avg_relevancy: z.number().nullable().optional(),
    avg_groundedness: z.number().nullable().optional(),
    avg_context_precision: z.number().nullable().optional(),
    avg_context_relevancy: z.number().nullable().optional(),
    avg_completeness: z.number().nullable().optional(),
    avg_correctness: z.number().nullable().optional(),
    avg_compliance_accuracy: z.number().nullable().optional(),
    avg_abstention: z.number().nullable().optional(),
    hallucination_rate: z.number().nullable().optional(),
    avg_chunk_alignment: z.number().nullable().optional(),
    profile_id: z.string().nullable().optional(),
    profile_version: z.string().nullable().optional(),
    judge_model_name: z.string().nullable().optional(),
    synthesis_model_name: z.string().nullable().optional(),
    retrieval_config: z.record(z.unknown()).nullable().optional(),
    corpus_snapshot: z.record(z.unknown()).nullable().optional(),
    overall_verdict: z.string().nullable().optional(),
    pass_count: z.number().nullable().optional(),
    fail_count: z.number().nullable().optional(),
    review_count: z.number().nullable().optional(),
    total_tokens: z.number().nullable().optional(),
    error_message: z.string().nullable().optional(),
    created_at: z.string().nullable().optional(),
    completed_at: z.string().nullable().optional(),
});

export const EvalRunDetailSchema = EvalRunSchema.extend({
    results: z.array(EvalResultSchema),
});

export const ComparisonMetricSchema = z.object({
    metric: z.string(),
    run_a: z.number().nullable().optional(),
    run_b: z.number().nullable().optional(),
    winner: z.string().nullable().optional(),
});

export const QuestionComparisonSchema = z.object({
    question: z.string(),
    expected_answer: z.string().nullable().optional(),
    run_a: EvalResultSchema.nullable().optional(),
    run_b: EvalResultSchema.nullable().optional(),
});

export const ComparisonDecisionSchema = z.object({
    winner: z.string().nullable().optional(),
    winner_name: z.string().nullable().optional(),
    decision_status: z.string(),
    reason_codes: z.array(z.string()),
    summary: z.string(),
    risk_flags: z.array(z.string()),
    disqualified: z.record(z.array(z.string())),
});

export const ComparisonWarningSchema = z.object({
    code: z.string(),
    message: z.string(),
});

export const ComparisonResponseSchema = z.object({
    run_a: EvalRunSchema,
    run_b: EvalRunSchema,
    metrics: z.array(ComparisonMetricSchema),
    questions: z.array(QuestionComparisonSchema),
    decision: ComparisonDecisionSchema.nullable().optional(),
    warnings: z.array(ComparisonWarningSchema).optional(),
});

export const SynthesizedQuestionSchema = z.object({
    question: z.string(),
    expected_answer: z.string().nullable().optional(),
    truth: TruthPayloadSchema.nullable().optional(),
});

export const SynthesizeResponseSchema = z.object({
    questions: z.array(SynthesizedQuestionSchema),
    count: z.number(),
});

export type EvalRunCreateResponse = z.infer<typeof EvalRunCreateResponseSchema>;
export type CoverageGaps = z.infer<typeof CoverageGapsSchema>;
export type DeterministicCheck = z.infer<typeof DeterministicCheckSchema>;
export type AnswerTruth = z.infer<typeof AnswerTruthSchema>;
export type RetrievalTruth = z.infer<typeof RetrievalTruthSchema>;
export type TruthMetadata = z.infer<typeof TruthMetadataSchema>;
export type TruthPayload = z.infer<typeof TruthPayloadSchema>;
export type EvalResult = z.infer<typeof EvalResultSchema>;
export type EvalRun = z.infer<typeof EvalRunSchema>;
export type EvalRunDetail = z.infer<typeof EvalRunDetailSchema>;
export type ComparisonMetric = z.infer<typeof ComparisonMetricSchema>;
export type QuestionComparison = z.infer<typeof QuestionComparisonSchema>;
export type ComparisonDecision = z.infer<typeof ComparisonDecisionSchema>;
export type ComparisonWarning = z.infer<typeof ComparisonWarningSchema>;
export type ComparisonResponse = z.infer<typeof ComparisonResponseSchema>;
export type SynthesizedQuestion = z.infer<typeof SynthesizedQuestionSchema>;
export type SynthesizeResponse = z.infer<typeof SynthesizeResponseSchema>;
